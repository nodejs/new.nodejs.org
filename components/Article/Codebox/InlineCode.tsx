import styles from './index.module.scss';
import type { FC, PropsWithChildren } from 'react';

const InlineCode: FC<PropsWithChildren> = ({ children }) => (
  <code className={styles.code}>{children}</code>
);

export default InlineCode;
