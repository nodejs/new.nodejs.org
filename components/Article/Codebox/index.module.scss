@use 'themes/light';
@use 'themes/dark';
@use 'styles/mixins/components';

.pre {
  @include components.code;

  .top {
    display: flex;
    flex-direction: row;
    justify-content: space-between;

    .lang,
    .copy {
      align-items: center;
      cursor: pointer;
      display: inherit;
      font-size: var(--font-size-code);
      height: 23px;
      justify-content: center;
      width: 86px;
    }

    .langBox {
      background-color: var(--black4);
      border-radius: 0 0 0.3rem 0;
      display: flex;
      flex-direction: row;
      justify-content: center;

      .lang {
        background-color: var(--black4);
        border-width: 0;
        color: var(--black9);
        padding: 0 16px;
        width: max-content;

        &:last-of-type {
          border-radius: 0 0 0.3rem 0;
        }

        &:hover {
          background-color: var(--black5);
        }
      }

      .lang.selected {
        font-weight: 600;
      }
    }

    .copy {
      background-color: var(--black9);
      border-radius: 0 0 0 0.3rem;
      border-width: 0;
      color: white;

      &:hover {
        background-color: var(--black8);
      }

      i {
        padding: 0;
      }
    }
  }

  .content {
    margin: 1em;
  }
}

[data-theme='light'] {
  .pre {
    @include light.theme;
  }
}

[data-theme='dark'] {
  .pre {
    @include dark.theme;

    .top {
      span {
        background-color: var(--black3);
        color: var(--black9);
      }

      .copy {
        background-color: var(--brand8);
        color: white;

        &:hover {
          background-color: var(--brand6);
        }
      }
    }
  }
}
