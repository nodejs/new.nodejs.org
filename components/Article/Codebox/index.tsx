import { useEffect, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import { highlight, languages } from 'prismjs';
import classnames from 'classnames';
import styles from './index.module.scss';
import { useCopyToClipboard } from '../../../hooks/useCopyToClipboard';
import type { FC, PropsWithChildren, ReactElement, MouseEvent } from 'react';

type CodeBoxProps = {
  children: ReactElement<PropsWithChildren<{ className?: string }>>;
};

export const replaceLabelLanguages = (language: string) =>
  language.replace(/console/i, 'bash');

export const replaceLanguages = (language: string) =>
  language
    .replace(/mjs|cjs|javascript/i, 'js')
    .replace(/console|shell/i, 'bash');

const Codebox: FC<CodeBoxProps> = ({ children: { props } }) => {
  const [parsedCode, setParsedCode] = useState('');
  const [copied, copyText] = useCopyToClipboard();
  const [langIndex, setLangIndex] = useState(0);

  const className = props.className || 'text';

  const languageOptions = className
    .split('|')
    .map(language => language.split('language-')[1]);

  const language = languageOptions[langIndex];

  const codeArray = props.children
    ? props.children.toString().split('--------------\n')
    : [''];

  const handleCopyCode = (event: MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    copyText(codeArray[langIndex]);
  };

  useEffect(() => {
    const parsedLanguage = replaceLanguages(language);
    const prismLanguage = languages[parsedLanguage] || languages.text;

    setParsedCode(
      highlight(codeArray[langIndex], prismLanguage, parsedLanguage)
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [langIndex]);

  return (
    <pre className={classnames(styles.pre, replaceLanguages(className))}>
      <div className={styles.top}>
        <div className={styles.langBox}>
          {languageOptions.map((lang, index) => (
            <button
              type="button"
              key={lang}
              className={classnames(styles.lang, {
                [styles.selected]: index === langIndex,
              })}
              onClick={() => setLangIndex(index)}
            >
              {replaceLabelLanguages(lang.toLowerCase())}
            </button>
          ))}
        </div>
        <button type="button" className={styles.copy} onClick={handleCopyCode}>
          <FormattedMessage id="components.codeBox.copy" values={{ copied }} />
        </button>
      </div>
      <div
        className={styles.content}
        dangerouslySetInnerHTML={{ __html: parsedCode }}
      />
    </pre>
  );
};

export default Codebox;
