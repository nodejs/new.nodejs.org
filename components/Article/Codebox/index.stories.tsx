import Codebox from './index';
import type { Meta as MetaObj, StoryObj } from '@storybook/react';
import type { FC } from 'react';

type DecoratedCodeBoxProps = { language: string[]; code: string[] };

const DecoratedCodeBox: FC<DecoratedCodeBoxProps> = ({ language, code }) => (
  <Codebox>
    <pre className={language.join('|')}>{code.join('--------------\n')}</pre>
  </Codebox>
);

type Story = StoryObj<typeof DecoratedCodeBox>;
type Meta = MetaObj<typeof DecoratedCodeBox>;

const singleLangCode = ['const a = 1;'];

export const Default: Story = {
  args: {
    language: ['language-js'],
    code: singleLangCode,
  },
};

const multiLangCode = [
  "const http = require('http');",
  "import http from 'http';",
];

export const MultiLang: Story = {
  args: {
    language: ['language-cjs', 'language-mjs'],
    code: multiLangCode,
  },
};

export default { component: DecoratedCodeBox } as Meta;
