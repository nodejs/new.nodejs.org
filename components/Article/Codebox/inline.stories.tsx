import InlineCode from './InlineCode';
import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof InlineCode> = {
  title: 'InlineCode',
  component: InlineCode,
};

export default meta;

type Story = StoryObj<typeof InlineCode>;

const code = 'const a = 1;';

export const Default: Story = {
  args: {
    children: <code>{code}</code>,
  },
};
