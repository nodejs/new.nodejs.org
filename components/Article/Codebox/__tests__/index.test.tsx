import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import Codebox, { replaceLabelLanguages, replaceLanguages } from '../index';

describe('Replacer tests', (): void => {
  it('replaceLabelLanguages', (): void => {
    expect(replaceLabelLanguages('language-console')).toBe('language-bash');
  });

  it('replaceLanguages', (): void => {
    expect(replaceLanguages('language-mjs')).toBe('language-js');
    expect(replaceLanguages('language-cjs')).toBe('language-js');
    expect(replaceLanguages('language-javascript')).toBe('language-js');
    expect(replaceLanguages('language-console')).toBe('language-bash');
    expect(replaceLanguages('language-shell')).toBe('language-bash');
  });
});

describe('Codebox component (one lang)', (): void => {
  const code = 'const a = 1;';

  it('should copy content', async () => {
    const user = userEvent.setup();

    render(
      <IntlProvider locale="en" onError={() => {}}>
        <Codebox>
          <pre className="language-js">{code}</pre>
        </Codebox>
      </IntlProvider>
    );

    const navigatorClipboardWriteTextSpy = jest.spyOn(
      navigator.clipboard,
      'writeText'
    );

    const buttonElement = screen.getByText('components.codeBox.copy');
    await user.click(buttonElement);

    expect(navigatorClipboardWriteTextSpy).toHaveBeenCalledTimes(1);
    expect(navigatorClipboardWriteTextSpy).toHaveBeenCalledWith(code);
  });
});

describe('Codebox component (multiple langs)', (): void => {
  const code = `const http = require('http');
--------------
import http from 'http';`;

  it('switch between languages', async () => {
    const user = userEvent.setup();

    const { container } = render(
      <IntlProvider locale="en" onError={() => {}}>
        <Codebox>
          <pre className="language-cjs|language-mjs">{code}</pre>
        </Codebox>
      </IntlProvider>
    );

    expect(container).toMatchSnapshot();

    const buttonElement = await screen.findByText('mjs');
    await user.click(buttonElement);

    expect(container).toMatchSnapshot();
  });
});
