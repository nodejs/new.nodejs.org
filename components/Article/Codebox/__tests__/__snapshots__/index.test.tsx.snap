// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Codebox component (multiple langs) switch between languages 1`] = `
<div>
  <pre
    class="pre language-js|language-mjs"
  >
    <div
      class="top"
    >
      <div
        class="langBox"
      >
        <button
          class="lang selected"
          type="button"
        >
          cjs
        </button>
        <button
          class="lang"
          type="button"
        >
          mjs
        </button>
      </div>
      <button
        class="copy"
        type="button"
      >
        components.codeBox.copy
      </button>
    </div>
    <div
      class="content"
    >
      <span
        class="token keyword"
      >
        const
      </span>
       http 
      <span
        class="token operator"
      >
        =
      </span>
       
      <span
        class="token function"
      >
        require
      </span>
      <span
        class="token punctuation"
      >
        (
      </span>
      <span
        class="token string"
      >
        'http'
      </span>
      <span
        class="token punctuation"
      >
        )
      </span>
      <span
        class="token punctuation"
      >
        ;
      </span>
      

    </div>
  </pre>
</div>
`;

exports[`Codebox component (multiple langs) switch between languages 2`] = `
<div>
  <pre
    class="pre language-js|language-mjs"
  >
    <div
      class="top"
    >
      <div
        class="langBox"
      >
        <button
          class="lang"
          type="button"
        >
          cjs
        </button>
        <button
          class="lang selected"
          type="button"
        >
          mjs
        </button>
      </div>
      <button
        class="copy"
        type="button"
      >
        components.codeBox.copy
      </button>
    </div>
    <div
      class="content"
    >
      <span
        class="token keyword"
      >
        import
      </span>
       http 
      <span
        class="token keyword"
      >
        from
      </span>
       
      <span
        class="token string"
      >
        'http'
      </span>
      <span
        class="token punctuation"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;
